log: Mitigate disk filling attacks by rate limiting LogPrintf, LogInfo, LogWarning, LogError

Crypt-iQ opened the pull request:
This revives the work done by dergoegge in https://github.com/bitcoin/bitcoin/pull/21603. The approach is similar — this PR uses `std::source_location` under the hood now that we can use c++20 features. It also resets the rate limiting statistics via the `CScheduler`. The logging functions have also changed slightly since that PR was opened, so work has been done to preserve the intent of the original rate limiting change. I have tried to give commit attribution where possible.

**Approach:**
Each source code location is given an hourly logging quota of 1MiB of logging per hour. Logging is only dropped from source locations that exceed the quota.
- Only logging to disk is rate limited. Logging to console is not rate limited.
- Printing with the category argument is not rate limited.
- `UpdateTip: new best=[…]` is logged without rate limiting. High log volume is expected for that source location during IBD.
- When logging is restarted a tally of how many bytes were dropped is printed.
- All logs will be prefixed with [*] if there is at least one source location that is currently being suppressed.

I've repurposed the old logging rpc mentioned in #21603 in another branch for testing [here](https://github.com/Crypt-iQ/bitcoin/tree/log_ratelimiting_05192025_rpc). This can be used to log from source locations and test out the new changes in logging behavior. Note that the `setmocktime` RPC needs to be used to set the mock time past the current clock time to reset the logging messages.

Example usage:
```
bitcoin-cli -regtest excessivelog 1 1048500 # log just under 1MiB
bitcoin-cli -regtest excessivelog 1 100 # this should get the total amount logged above 1MiB
                                        # and the rate limiting logic should kick in
bitcoin-cli -regtest excessivelog 2 1048500
bitcoin-cli -regtest excessivelog 2 100 # trigger rate limiting from another location
bitcoin-cli -regtest mockscheduler 3600 # fast-forward the scheduler
bitcoin-cli -regtest excessivelog 1 100 # this should trigger logging to resume and will log the source locations that were reset
```

Example output:
```
2025-07-02T22:03:56Z [warning] Excessive logging detected from rpc/node.cpp:142 (RPCHelpMan excessivelog()): >1048576 bytes logged during the last time window of 3600s. Suppressing logging to disk from this source location until time window resets. Console logging unaffected. Last log entry.
[*] 2025-07-02T22:03:56Z aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
2025-07-02T22:04:58Z (mocktime: 2262-04-11T23:47:15Z) Restarting logging from rpc/node.cpp:142 (RPCHelpMan excessivelog()): 121 bytes were dropped during the last 3600s.
2025-07-02T22:04:58Z (mocktime: 2262-04-11T23:47:15Z) Restarting logging from rpc/node.cpp:139 (RPCHelpMan excessivelog()): 121 bytes were dropped during the last 3600s.
```

davidgumberg commented:
Concept ACK

dergoegge commented:
Concept ACK

Thanks for picking this up!

janb84 commented on src/logging.h:
```diff
@@ -27,10 +31,29 @@ static const bool DEFAULT_LOGTIMESTAMPS = true;
 static const bool DEFAULT_LOGTHREADNAMES = false;
 static const bool DEFAULT_LOGSOURCELOCATIONS = false;
 static constexpr bool DEFAULT_LOGLEVELALWAYS = false;
```
nit: would be nice to align the other initializations to the new C++ 11 {} style in a follow up PR, to keep the code base consistent.

Crypt-iQ replied:
I have opted not to do this in the follow-up PR.

janb84 reviewed (commented):
ACK 0bc6ed61cfab6d97e74103efd41c46faf5941ff6

Ratelimits loging to disk if loggin exceeds 1 MiB  in  1 hour (WINDOW_MAX_BYTES const & WINDOW_SIZE const)
 
- code review 
- build & tested

brunoerg commented:
Concept ACK

Crypt-iQ commented:
The tidy job is failing because it doesn't like the logging jobs used in lambdas. It seems like this is pre-existing so I'm not sure why it's failing now. The windows cross-built job is failing on the `rate_limiting` test at every file-size comparison. I don't have a windows machine to debug this, but I think maybe `fs::file_size` is failing or some other quirk is showing up?

stickies-v commented on src/logging.h:
```diff
@@ -113,17 +181,17 @@ namespace BCLog {
 
         /** Log categories bitfield. */
-        std::atomic<CategoryMask> m_categories{BCLog::NONE};
+        std::atomic<CategoryMask> m_categories{DEFAULT_LOG_FLAGS};
```
Hmm, I'm feeling uneasy about using `m_categories`, when rate limiting flags fundamentally don't represent a category. I can see how it reduces the diff, but I'm not sure that's worth it. I'll think about an alternative approach - have you considered any already?

Since the user doesn't have to be able to configure any of this, and we only have a single `UNCONDITIONAL_ALWAYS` use case (and likely won't have (many) more in the future), I think adding an `is_conditional=True` arg to `LogPrintLevel` and a `LogInfoAlways` macro might be an alternative?

stickies-v replied:
Or perhaps even better than a boolean is a `uint64_t rate_limit=DEFAULT_RATE_LIMIT` so we can do away with the distinction between conditional and unconditional logging entirely? We could use rate_limit=0 as a synonym for unconditional logging, but I think even `UpdateTipLog()` doesn't have to be exempt from rate limiting, we could still set a sane value that allows doing entire IBD in one hour, for example.

Then `UpdateTipLog()` could just use:
```
    LogInfo(/*rate_limit=*/100, "%s%s: new best=%s height=%d version=0x%08x log2_work=%f tx=%lu date='%s' progress=%f cache=%.1fMiB(%utxo)%s\n",
```

Crypt-iQ replied:
I came to the conclusion that because macros don't have default arguments, skipping the rate limiting for `UpdateTipLog` means that we need to: 1) introduce a new macro for IBD specifically for `UpdateTipLog`, or 2) modify all call-sites of whichever macro `UpdateTipLog` uses to accept a `rate_limit` argument.

I don't want to introduce a new macro only for tip logging since there are quite a few already, but I think it could be a better choice than adding more noise to the categories. I also think that suggestion 2) to modify all call-sites to pass in a rate-limiting argument is cumbersome when most callers don't need to be aware of rate limiting and it should instead be done under-the-hood (i.e. `UpdateTipLog` is the exception to the rule).

I might be wrong here, curious to hear thoughts.

stickies-v replied:
> 2) to modify all call-sites to pass in a rate-limiting argument is cumbersome

Absolutely agree that would not be an improvement.

> I think using a boolean for rate-limiting vs no-rate-limiting is better than a uint64_t

I think either will work fine, the uint64_t generalizes a bit nicer but we probably won't be needing the generalization in the future, so whichever ends up being the most elegant implementation now is what I think we should go for.

> I am partial to introducing a new macro for `UpdateTipLog`

That seems sensible to me, yes.

Crypt-iQ replied:
I've changed this so that the logging macros eventually pass a `should_ratelimit` bool to `LogPrintStr`. Instead of introducing a new macro for `UpdateTipLog`, it now uses `LogPrintLevel` which can bypass the rate-limiting. I'm not sure how I feel about using `LogPrintLevel` as I don't want to encourage more usage of it in the codebase as it could be a potential footgun.

stickies-v reviewed (commented):
Concept ACK on locking down this attack vector, and the code changes required seem acceptable. Also nice to start using `std::source_location` here.

Crypt-iQ commented on src/init/common.cpp:
```diff
@@ -40,6 +40,7 @@ void AddLoggingArgs(ArgsManager& argsman)
     argsman.AddArg("-loglevelalways", strprintf("Always prepend a category and level (default: %u)", DEFAULT_LOGLEVELALWAYS), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-printtoconsole", "Send trace/debug info to console (default: 1 when no -daemon. To disable logging to file, set -nodebuglogfile)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-shrinkdebugfile", "Shrink debug.log file on client startup (default: 1 when no -debug)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-ratelimitlogging", strprintf("Rate limit unconditional logging to disk (default: %u)", DEFAULT_RATELIMITLOGGING), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
That's a good point, I can't think of any concrete use case to disable rate limiting here.

fanquake replied:
> Since the attack only matters when peers can influence us, could we simply disable rate-limiting for all logs during IBD

?

l0rinc replied:
Can you be more specific :)?

l0rinc replied:
Thanks for clarifying

stickies-v commented on src/logging.cpp:
```diff
@@ -364,33 +368,71 @@ std::string BCLog::Logger::GetLogPrefix(BCLog::LogFlags category, BCLog::Level l
 }
 
-void BCLog::Logger::FormatLogStrInPlace(std::string& str, BCLog::LogFlags category, BCLog::Level level, std::string_view source_file, int source_line, std::string_view logging_function, std::string_view threadname, SystemClock::time_point now, std::chrono::seconds mocktime) const
+bool BCLog::Logger::FormatLogStrAndRateLimit(std::string& str, BCLog::LogFlags category, BCLog::Level level, const std::source_location& source_loc, std::string_view logging_function, std::string_view threadname, SystemClock::time_point now, std::chrono::seconds mocktime)
```
I think it's unintuitive that this function returns `true` when the operation failed. It seems like the rate limiting can be carved out pretty much as-is into a separate `bool NeedsRateLimiting()` function? There are only 2 callsites of `FormatLogStrAndRateLimit`, and the first one (`StartLogging`) doesn't look like it needs rate limiting because we already have a max buffer size?

Crypt-iQ replied:
I've carved this out into a `NeedsRateLimiting` function and have kept the logic of `FormatLogStrInPlace` mostly unchanged. The reason I had the original change is because `FormatLogStrInPlace` inserts a timestamp at the beginning of `str` and I wanted the prepended rate-limiting logs to occur _before_ the timestamp. With the new change, the rate-limiting logs are first in `str` followed by the time-stamp. I think now the logs look a little wonky, but I like the code change better. I could also extract the time-stamp prepend logic so it's called after `NeedsRateLimiting`.

hebasto commented:
> > The windows cross-built job is failing on the rate_limiting test at every file-size comparison. I don't have a windows machine to debug this, but I think maybe fs::file_size is failing or some other quirk is showing up?
> 
> @hebasto can you take a look here? I ran `test_bitcoin.exe` from this branch (rebased) under wine and didn't see any failures.

It seems to be related to MSVCRT behaviour, as binaries [linked to UCRT](https://github.com/hebasto/bitcoin-core-nightly/actions/runs/15345658695) pass all tests.

stickies-v commented on src/logging.h:
```diff
@@ -81,16 +106,54 @@ namespace BCLog {
+    {
+    private:
+        //! Timestamp of the last window reset.
+        std::chrono::time_point<NodeClock> m_last_reset;
```
Not a big issue, but since we have quite a few log source locations, this class will be instantiated a fair amount of (`N`) times. An alternative approach would be to have a single `LogRateLimiter` instance that encapsulates most of the logic in 6b54362c19ca9baf9dfa9be9281f6faeda169420, and has a single `m_last_reset` member, spawning a minimal `SourceLocationCounter` struct for each source location, containing just the `m_available_bytes` and `m_dropped_bytes` members.

This:
- reduces `N` `std::chrono::time_point<NodeClock>` instances to just 1
- allows `LogRateLimiter` to completely clean up memory for log locations that haven't been triggered in the last hour
- improves code encapsulation

Overall, this shouldn't be a big change from what you have now, mostly just moving code around into a new class?

Crypt-iQ replied:
Implemented this change, lmk what you think. I think this encapsulates code a bit better, but it probably could still be improved.

stickies-v commented on src/logging.cpp:
```diff
@@ -364,33 +368,71 @@ std::string BCLog::Logger::GetLogPrefix(BCLog::LogFlags category, BCLog::Level l
+                                    "(%d MiB) were dropped during the last hour.\n",
+                                    source_loc.file_name(), source_loc.line(), logging_function,
+                                    dropped_bytes / (1024 * 1024)));
+        } else if (is_ratelimited && !was_ratelimited) {
```
nit: It would be nice if we could catch this in debug builds (ideally even at a fraction, say 20% of the limit), to help inform if our default limit needs to change.  `Assume()` is an option, but that would probably break unit tests in debug builds.

nit, because 1MB is small enough that it should prevent issues even on devices with little disk size (unless an attacker finds multiple log vulnerabilities), and large enough that no single unconditional logging statement should ever hit that in the foreseeable future, so I think "driving blind" is acceptable here.

Crypt-iQ replied:
I want to address this because I think it would be useful to catch in debug builds via `Assume` but also don't want to break tests. Do you think this should be addressed?

dergoegge commented:
I vaguely remember having trouble with the unit tests on windows too. Iirc it had something to do with windows line endings being different (i.e. `\r\n` on windows vs `\n` on linux), which throws off the size accounting.

Crypt-iQ commented on src/logging.h:
```diff
@@ -105,6 +168,11 @@ namespace BCLog {
         size_t m_buffer_lines_discarded GUARDED_BY(m_cs){0};
 
+        //! Fixed window rate limiters for each source location that has attempted to log something.
+        std::unordered_map<std::source_location, LogRateLimiter, SourceLocationHasher, SourceLocationEqual> m_ratelimiters GUARDED_BY(m_cs);
```
This never gets cleared. I haven't calculated the overhead of the map, but it could make sense to clear it periodically.

stickies-v replied:
note: this is part of my suggestion in https://github.com/bitcoin/bitcoin/pull/32604#discussion_r2115841360

l0rinc replied:
@stickies-v, my mistake, I quickly wanted to make sure this doesn't affect my IBD benchmarks, didn't have time to review it in detail yet

Crypt-iQ replied:
Both the map and set are now cleared if `LogRateLimiter` determines that the logging window should be reset.

Crypt-iQ commented on src/logging.h:
```diff
@@ -39,39 +62,41 @@ struct LogCategory {
+        SCAN                       = (CategoryMask{1} << 27),
+        TXPACKAGES                 = (CategoryMask{1} << 28),
+        UNCONDITIONAL_RATE_LIMITED = (CategoryMask{1} << 29),
+        UNCONDITIONAL_ALWAYS       = (CategoryMask{1} << 30),
```
Resolving as I've removed categories.

Crypt-iQ reviewed (commented):
The latest change uses a single `BCLog::LogRateLimiter` instance and uses a slimmer way of tracking the amount of logging bytes left per `std::source_location`. I like the approach better as `m_categories` is no longer being used. I would like to point out that the rate-limiting logic can be bypassed if `LogPrintLevel` is used. I think `LogPrintLevel` callers need to be cautious and I kind of think this macro shouldn't be used if `level >= BCLog::Level::Info`.

> Will that appear on the console as well, which didn't have any messages dropped?

Yes this appears in the console as well. I actually was unsure if console-logging should be rate-limited or not. Hopefully users have sensible limits on their terminal history?

I will work on fixing the CI, I think review can still happen in the meantime while I debug the windows failure and fix the tidy job.

Crypt-iQ committed:
log: introduce LogRateLimiter, LogLimitStats, Status

LogRateLimiter will be used to keep track of source locations and our
current time-based logging window. It contains an unordered_map and a
m_suppressions_active bool to track source locations. The map is keyed
by std::source_location, so a custom Hash function (SourceLocationHasher)
and custom KeyEqual function (SourceLocationEqual) is provided.
SourceLocationHasher uses CSipHasher(0,0) under the hood to get a
uniform distribution.

A public Reset method is provided so that a scheduler (e.g. the
"b-scheduler" thread) can periodically reset LogRateLimiter's state when
the time window has elapsed.

The LogRateLimiter::Consume method checks if we have enough available
bytes in our rate limiting budget to log an additional string. It
returns a Status enum that denotes the rate limiting status and can
be used by the caller to emit a warning, skip logging, etc.

The Status enum has three states:
- UNSUPPRESSED     (logging was successful)
- NEWLY_SUPPRESSED (logging was succcesful, next log will be suppressed)
- STILL_SUPPRESSED (logging was unsuccessful)

LogLimitStats counts the available bytes left for logging per source
location for the current logging window. It does not track actual source
locations; it is used as a value in m_source_locations.

Also exposes a SuppressionsActive() method so the logger can use
that in a later commit to prefix [*] to logs whenenever suppressions
are active.

Co-Authored-By: Niklas Gogge <n.goeggi@gmail.com>
Co-Authored-By: stickies-v <stickies-v@protonmail.com>

Crypt-iQ committed:
log: use std::source_location in place of __func__, __FILE__, __LINE__

The std::source_location conveniently stores the file name, line number,
and function name of a source code location. We switch to using it instead
of the __func__ identifier and the __FILE__ and __LINE__ macros.

BufferedLog is changed to have a std::source_location member, replacing the
source_file, source_line, and logging_function members. As a result,
MemUsage no longer explicitly counts source_file or logging_function as the
std::source_location memory usage is included in the MallocUsage call.

This also changes the behavior of -logsourcelocations as std::source_location
includes the entire function signature. Because of this, the functional test
feature_config_args.py must be changed to no longer include the function
signature as the function signature can differ across platforms.

Co-Authored-By: Niklas Gogge <n.goeggi@gmail.com>
Co-Authored-By: stickies-v <stickies-v@protonmail.com>

Crypt-iQ committed:
log: Add rate limiting to LogPrintf, LogInfo, LogWarning, LogError, LogPrintLevel

To mitigate disk-filling attacks caused by unsafe usages of LogPrintf and
friends, we rate-limit them by passing a should_ratelimit bool that
eventually makes its way to LogPrintStr which may call
LogRateLimiter::Consume. The rate limiting is accomplished by
adding a LogRateLimiter member to BCLog::Logger which tracks source
code locations for the given logging window.

Every hour, a source location can log up to 1MiB of data. Source
locations that exceed the limit will have their logs suppressed for the
rest of the window determined by m_limiter.

This change affects the public LogPrintLevel function if called with
a level >= BCLog::Level::Info.

The UpdateTipLog function has been changed to use the private LogPrintLevel_
macro with should_ratelimit set to false. This allows UpdateTipLog to log
during IBD without hitting the rate limit.

Note that on restart, a source location that was rate limited before the
restart will be able to log until it hits the rate limit again.

Co-Authored-By: Niklas Gogge <n.goeggi@gmail.com>
Co-Authored-By: stickies-v <stickies-v@protonmail.com>

Crypt-iQ committed:
test: Mark ~DebugLogHelper as noexcept(false)

We mark ~DebugLogHelper as noexcept(false) to be able to catch the
exception it throws. This lets us use it in test in combination with
BOOST_CHECK_THROW and BOOST_CHECK_NO_THROW to check that certain log
messages are (not) logged.

Co-Authored-By: Niklas Gogge <n.goeggi@gmail.com>

Crypt-iQ commented on src/logging.h:
```diff
@@ -269,11 +353,14 @@ inline void LogPrintFormatInternal(std::string_view logging_function, std::strin
+// Note that conditional logging is performed WITHOUT rate limiting. Users specifying
+// -debug are assumed to be developers or power users who are aware that -debug may cause
+// excessive disk usage due to logging.
+#define LogPrintLevel(category, level, ...)                                           \
```
I've changed the misleading docstring and I've implemented suggestion 1. and I will pick up suggestion 2. as well. I think this might conflict a bit with the work in https://github.com/bitcoin/bitcoin/pull/29256.

stickies-v replied:
Nice. Since af86b55f996cf06e8e9179d5c1fe7ce3d94fa3c6 now uses `LogPrintLevel_` directly, I think the changes to `LogPrintLevel` in ed3c0a592b60a565ec5dade7c3af62e7f4cf75a1 can now be reverted and rate limiting applied consistently to all "public" macros? (would also require an update to the `UpdateTipLog` docstring)

stickies-v replied:
Oh no I agree, debug logs should be exempt from ratelimiting. My point is just that I think it's bad to have different logging behaviour depending on whether `LogInfo()` or `LogPrintLevel(..., BCLog::Info)` is used, so I think something like the below would work well?

<details>
<summary>git diff on 911ee520c8</summary>

```diff
diff --git a/src/logging.h b/src/logging.h
index d588ef86dc..5f298405e7 100644
--- a/src/logging.h
+++ b/src/logging.h
@@ -349,17 +349,13 @@ inline void LogPrintFormatInternal(const std::source_location& source_loc, const
 // Use a macro instead of a function for conditional logging to prevent
 // evaluating arguments when logging for the category is not enabled.
 
-// Log by prefixing the output with the passed category name and severity level. This can either
-// log conditionally if the category is allowed or unconditionally if level >= BCLog::Level::Info
-// is passed. Note that this function bypasses rate limiting. Callers need to take care to ensure
-// that an attacker cannot trigger a disk-fill vulnerability if level >= Info is used. Additionally,
-// users specifying -debug are assumed to be developers or power users who are aware that -debug
-// may cause excessive disk usage due to logging.
-#define LogPrintLevel(category, level, ...)                                           \
-    do {                                                                              \
-        if (LogAcceptCategory((category), (level))) {                                 \
-            LogPrintLevel_(category, level, /*should_ratelimit=*/false, __VA_ARGS__); \
-        }                                                                             \
+// Log conditionally, prefixing the output with the passed category name and severity level.
+#define LogPrintLevel(category, level, ...)                             \
+    do {                                                                \
+        if (LogAcceptCategory((category), (level))) {                   \
+            bool rate_limit{level >= BCLog::Info};                      \
+            LogPrintLevel_(category, level, rate_limit, __VA_ARGS__);   \
+        }                                                               \
     } while (0)
 
 // Log conditionally, prefixing the output with the passed category name.

```
</details>

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -364,17 +373,63 @@ std::string BCLog::Logger::GetLogPrefix(BCLog::LogFlags category, BCLog::Level l
 }
 
-void BCLog::Logger::FormatLogStrInPlace(std::string& str, BCLog::LogFlags category, BCLog::Level level, std::string_view source_file, int source_line, std::string_view logging_function, std::string_view threadname, SystemClock::time_point now, std::chrono::seconds mocktime) const
+bool BCLog::Logger::NeedsRateLimiting(bool should_ratelimit, const std::source_location& source_loc, std::string& str, std::string_view logging_function)
```
I've removed `logging_function` and `__func__` usage. I think it makes the function names a little longer in the map, but I'm ok with it. I think `m_cs` is unnecessary if `m_ratelimiter` is guarded by `m_cs`?

Crypt-iQ replied:
> I don't see how this affects any maps, though? I think it just affects the messages when we stop and restart logging?

Oh, I think I mistakenly thought that `function_name` was included in the map somehow or used when hashing `std::source_location`, but I see that's not actually the case.

Crypt-iQ commented:
I think the windows cross-built job is failing because a static BCLog::Logger* is used and is triggering the rate-limiting logic before the logging tests are run.

The [CI logs](https://github.com/bitcoin/bitcoin/actions/runs/15475499996/job/43571253106?pr=32604: 2025-06-05T19:56:44.9566257Z) have an asterisk:
`Mismatch at position 0: [*] foo5: bar5 != foo5: bar5`

stickies-v reviewed (commented):
Approach ACK. I think this PR needs release notes because of the behaviour change.

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -400,10 +450,8 @@ void BCLog::Logger::LogPrintStr_(std::string_view str, std::string_view logging_
-                .source_file=std::string(source_file),
                 .threadname=util::ThreadGetInternalName(),
-                .source_line=source_line,
+                .source_loc=source_loc,
```
I've sort of implemented this. I wasn't able to use `std::source_location&&` in `LogPrintStr_` because it's passed to two functions (`FormatLogStrInPlace` and `NeedsRateLimiting`) and implicit invalidation might occur. Instead, I pass `std::source_location&` and call `std::move` only when needed to invoke the move constructor.

Crypt-iQ replied:
I misunderstood the original suggestion and have now implemented it in ca0c1a7a9ce9198c950e123d36c9e4bb13db13da

stickies-v reviewed (approved):
utACK b7d9c669cadf666f595a3e20ee3726a9ede88acd modulo [missing release notes](https://github.com/bitcoin/bitcoin/pull/32604/files#r2131883418)

Thanks for quickly addressing my copious amounts of comments, it's been fun reviewing this. I've left a few (final?) nits, but none are blocking. Would prefer to squash the last 2 test commits into the 2 commits where the relevant functionality was introduced, but either is fine.

Crypt-iQ commented on src/validation.cpp:
```diff
@@ -2976,15 +2976,19 @@ static void UpdateTipLog(
-        coins_tip.GetCacheSize(),
-        !warning_messages.empty() ? strprintf(" warning='%s'", warning_messages) : "");
+
+    // We use the private LogPrintLevel_ macro here because we want to disable rate limiting from this location. This code
+    // could use LogPrintLevel with the Info level to bypass rate-limiting, but it is more opaque. Using LogPrintLevel_
```
I think this is an old comment, will change.

stickies-v commented on src/logging.cpp:
```diff
@@ -364,7 +365,7 @@ std::string BCLog::Logger::GetLogPrefix(BCLog::LogFlags category, BCLog::Level l
 static size_t MemUsage(const BCLog::Logger::BufferedLog& buflog)
 {
-    return buflog.str.size() + buflog.logging_function.size() + buflog.source_file.size() + buflog.threadname.size() + memusage::MallocUsage(sizeof(memusage::list_node<BCLog::Logger::BufferedLog>));
+    return buflog.str.size() + strlen(buflog.source_loc.function_name()) + strlen(buflog.source_loc.file_name()) + buflog.threadname.size() + memusage::MallocUsage(sizeof(memusage::list_node<BCLog::Logger::BufferedLog>));
```
I was long overdue for my biannual confusion about (dynamic) memory usage, but luckily here it is. It seems like I got it entirely backwards in my previous comment. The memory pointed to by `std::source_location` is static, so I don't think we need to account for any of it beyond what the `memusage::MallocUsage(sizeof(memusage::list_node<BCLog::Logger::BufferedLog>))` statement is already doing.

Orthogonal, but since we're touching the line anyway: dynamic memory usage of `std::string` should be measured with `.capacity()` instead of `.size()`.  Or even better, use the [`DynamicUsage(const std::string&)`](https://github.com/bitcoin/bitcoin/blob/154b98a7aaae248f5d40ca567d1ec51410d09bae/src/memusage.h#L94) function we already have to also account for small string optimization, which, of course, is totally a thing.

So the function could become:

```cpp
static size_t MemUsage(const BCLog::Logger::BufferedLog& buflog)
{
    return memusage::DynamicUsage(buflog.str) +
           memusage::DynamicUsage(buflog.threadname) +
           memusage::MallocUsage(sizeof(memusage::list_node<BCLog::Logger::BufferedLog>));
}
```

All in all, this should mean that thanks to ca0c1a7a9ce9198c950e123d36c9e4bb13db13da we can now buffer quite a bit more messages within our 1MB limit. Obviously not going to change anything meaningful, but wouldn't hurt to add a line to commit msg to highlight that.

Crypt-iQ replied:
Nice catch, I missed that `std::source_location` was part of `BufferedLog` and therefore included in the `MallocUsage` call.

l0rinc replied:
👍

Crypt-iQ committed:
doc: add release notes for new rate limiting logging behavior

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -364,17 +373,58 @@ std::string BCLog::Logger::GetLogPrefix(BCLog::LogFlags category, BCLog::Level l
+        str.insert(0, strprintf("Restarting logging from %s:%d (%s): "
+                                "(%d MiB) were dropped during the last hour.\n",
+                                source_loc.file_name(), source_loc.line(), source_loc.function_name(),
+                                dropped_bytes / (1024 * 1024)));
```
Deciding to keep as-is

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -364,17 +373,58 @@ std::string BCLog::Logger::GetLogPrefix(BCLog::LogFlags category, BCLog::Level l
 
     if (m_log_sourcelocations) {
-        str.insert(0, strprintf("[%s:%d] [%s] ", RemovePrefixView(source_file, "./"), source_line, logging_function));
+        str.insert(0, strprintf("[%s:%d] [%s] ", RemovePrefixView(source_loc.file_name(), "./"), source_loc.line(), source_loc.function_name()));
```
Done

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -86,24 +95,34 @@ BOOST_AUTO_TEST_CASE(logging_timer)
+        "[" + file_name + ":" + util::ToString(source_locs[2].line()) + "] [" + function_name + "] [debug] foo3: bar3",
+        "[" + file_name + ":" + util::ToString(source_locs[3].line()) + "] [" + function_name + "] foo4: bar4",
+        "[" + file_name + ":" + util::ToString(source_locs[4].line()) + "] [" + function_name + "] [debug] foo5: bar5",
+        "[" + file_name + ":" + util::ToString(source_locs[5].line()) + "] [" + function_name + "] foo6: bar6",
```
done

l0rinc reviewed (approved):
Can you please update the PR descripton now that there's no `UNCONDITIONAL_ALWAYS category`?

I'm mostly ok with the change, I'm testing if it introduces any slowdown and would prefer making the tests a bit more typesafe and compact (and the failure seems related to the change) and some of the new classes could use a more descriptive naming in my opinion.

stickies-v commented on src/logging.h:
```diff
@@ -31,6 +35,24 @@ extern const char * const DEFAULT_DEBUGLOGFILE;
+struct SourceLocationHasher {
+    size_t operator()(const std::source_location& s) const noexcept
+    {
+        // Use CSipHasher(0, 0) as a simple way to get uniform distribution.
```
I don't think this is a dead comment, the choice for using siphash over e.g. std::hash is not obvious (to me), and just naming the variable `uniform` would not have been clear to me. Would prefer keeping this as is.

l0rinc replied:
It's possible that this isn't the place, but it doesn't currently explain performance or collision resistance either - but if those are importance, maybe we could explain them in the commit message instead.

Crypt-iQ replied:
I have kept the comment -- I made a note in the commit message about the change.

l0rinc commented on src/logging.cpp:
Misunderstood that part, thanks. The different base part still applies (though I know this isn't uncommon in the code base)

Crypt-iQ replied:
I agree this is confusing and threw me. The different base part is pre-existing so will resolve. I think another PR could eliminate the different bases unless there's a compelling "historical" reason to keep both?

stickies-v commented on src/logging.h:
```diff
@@ -82,6 +104,65 @@ namespace BCLog {
     };
     constexpr auto DEFAULT_LOG_LEVEL{Level::Debug};
     constexpr size_t DEFAULT_MAX_LOG_BUFFER{1'000'000}; // buffer up to 1MB of log data prior to StartLogging
+    constexpr uint64_t RATELIMIT_MAX_BYTES{1024 * 1024}; // maximum number of bytes that can be logged within one window
```
Both are frequently used throughout the codebase. `RATELIMIT_MAX_BYTES` and `DEFAULT_MAX_LOG_BUFFER` are not related to eachother beyond both being used in logging. I don't see any footgun potential between using one or the other, so this seems like an either-is-fine option that can be left to the author to decide, and very quickly becomes bikeshedding otherwise.

stickies-v commented on src/logging.h:
```diff
@@ -82,15 +104,74 @@ namespace BCLog {
     };
     constexpr auto DEFAULT_LOG_LEVEL{Level::Debug};
```
This variable is not used or changed in this PR.

l0rinc replied:
This argument may not be, but the test is validating the changed code, dead parameter names don't increase my confidence in the correctness of the PR. It may have an obvious explanation, but it seemed related to me, given that we're touching surrounding code

stickies-v replied:
What do you mean with "dead parameter names"? This variable defines the default log level, which indeed is debug level. If you run `bitcoind -debug=net`, you'll start seeing net debug logs, you don't have to set `-loglevel=debug`.

l0rinc replied:
I replied to the wrong thread here, thanks for the explanation, please resolve it.

stickies-v commented on src/test/logging_tests.cpp:
```diff
@@ -60,6 +67,9 @@ struct LogSetup : public BasicTestingSetup {
         LogInstance().SetLogLevel(BCLog::Level::Debug);
         LogInstance().SetCategoryLogLevel({});
+
+        // Prevent tests from getting rate-limited.
```
The latter. "Ensure tests don't use each other's rate limiting budget" could be an alternative?

stickies-v commented on src/logging.h:
```diff
@@ -248,32 +335,38 @@ inline void LogPrintFormatInternal(std::string_view logging_function, std::strin
+#define LogPrintLevel_(category, level, should_ratelimit, ...) LogPrintFormatInternal(std::source_location::current(), category, level, should_ratelimit, __VA_ARGS__)
 
-// Log unconditionally.
+// Log unconditionally. Uses basic rate limiting to mitigate disk filling attacks.
```
Semantically, I agree. But practically, I think it's fine as is. Conditional vs unconditional are at this point fairly established concepts in our logging code in that they indicate whether or not logging categories need to be enabled or not. Rate limiting should in practice never occur, so for all intents and purposes, it is indeed unconditional.

How would you suggest to improve this? I can't think of an improvement.

l0rinc replied:
I don't consider "established" a relevant argument in this case - it just prevents new contributors from understanding the codebase. But if you think this isn't confusing, I don't mind resolving the comment, I did eventually understand it, just got sidetracked by the contradiction and wanted to simplify it to the next person. If we do correct it, we could be more specific instead of overly generalizing it (ie specify the restrictions instead of claiming there are none).

l0rinc commented on src/logging.cpp:
```diff
@@ -457,6 +466,7 @@ void BCLog::Logger::LogPrintStr_(std::string_view str, std::source_location&& so
     }
 
     FormatLogStrInPlace(str_prefixed, category, level, source_loc, util::ThreadGetInternalName(), SystemClock::now(), GetMockTime());
+    bool ratelimit = should_ratelimit && m_limiter.NeedsRateLimiting(source_loc, str_prefixed);
```
(I'm on a phone now, can't check the source)
Ah, so there's a side-effect here that wasn't obvious to me, thanks for clarifying.

stickies-v replied:
I agree `NeedsRateLimiting` has too many responsibilities. I think it's okay as-is, but your comment nerdsniped me into revisiting this. I'm working on a branch that removes the side effects from `NeedsRateLimiting` and a couple of other improvements. I'm not sure yet if it'll be too much of a detour for this PR, but at least it'll be interesting to be able to compare. Should have it ready tomorrow.

stickies-v replied:
Sorry for the wait, branch is ready: https://github.com/stickies-v/bitcoin/commits/2025-06/schedule-ratelimit-reset/ . I added the "fixup" comments as commits that should be squashable into the current branch as-is (requiring rebase for the commits above, though). Hopefully the isolation of the changes makes this easier to review and potentially adopt into this PR.

The main change is that I've carved out scheduling from `LogRateLimiter` (leaving that to `CScheduler`), which allows quite a bit of subsequent improvements. Specifically for this thread, I've [refactored](https://github.com/stickies-v/bitcoin/commit/c466e7b8ef11d0feb762c2b2ed41db695cd21602) `NeedsRateLimiting` into `Consume` that does not have any side-effects (except update internal accounting, as expected by the Consume name).

The other changes (see [diff](https://github.com/Crypt-iQ/bitcoin/compare/2ac8696b53e455dd27c8341828404a23b5cb68a9..9fb36ba46bccad079b4aea6062f7d187b7ef7775)) I've implemented in this branch include:
- change `Logger::m_limiter` to a pointer so it can more easily be (re)set, allowing for easier testing.
- add unit tests specifically for the `LogRateLimiter` interface that I find quite a bit easier to understand than the existing `rate_limit` tests, but there may be merit in keeping both (I've updated the latter so they don't break, but they should probably be deduplicated at least somewhat)
- made `LogRateLimiter` thread-safe since it is now used in multiple threads
- various cleanups, including removing the unnecessary `m_suppressed_locations` (now a `m_suppression_active` boolean), parameterizing the max_bytes and reset_window, ...
- removes the need for the test-only `BCLog::Logger::ResetLimiter()` function

I think the PR is in a good enough shape as it is, and I don't want to block progress. I like the new interface of my suggestion quite a bit better than the old one, and I think it is worth the time and effort to overhaul this (I spent a lot of time carving everything out in separate commits to make rebasing easy), but I understand this PR has had a lot of back-and-forth already. I'll leave it up to the author if he wants to change the approach here, of if I'll do it in a follow-up (which unfortunately would be a lot of churn). In either case, I'll shortly do a full re-review.

Crypt-iQ replied:
Latest change address this.

dergoegge reviewed (changes requested):
Code review ACK 2ac8696b53e455dd27c8341828404a23b5cb68a9

I think the commit order needs to change, so that 3c30ee9107fd0e916f9784b091a4d02f3a73ce46 comes before the tests that make use of the change?

nit: I'd also suggest to squash b44d31455ad46ca3ed95690dfc0a913445b9c1c9 and 5fe08e01384c4a4c8525060d365fa38c312758f0 into one commit, so there is no intermediate state between the commits where the validation logs are rate limited.

stickies-v reviewed (commented):
This [thread](https://github.com/bitcoin/bitcoin/pull/32604/files#r2164992810) nerdsniped me into reimplementing this PR with a different approach, leaving the scheduled resetting to `CScheduler`. This feels like a much more natural fit to me, and allows a couple of other improvements. See branch [here](https://github.com/stickies-v/bitcoin/commits/2025-06/schedule-ratelimit-reset/), or diff [here](https://github.com/Crypt-iQ/bitcoin/compare/2ac8696b53e455dd27c8341828404a23b5cb68a9..9fb36ba46bccad079b4aea6062f7d187b7ef7775).

Either way, I'll do a full re-review soon. I think the PR achieves its (important) stated goals, and while I think my suggested improvement is worthwhile, it isn't necessary or blocking. Making progress is more important than getting it perfectly right here.

stickies-v reviewed (approved):
ACK 2ac8696b53e455dd27c8341828404a23b5cb68a9

I still would prefer the approach I suggested [here](https://github.com/bitcoin/bitcoin/pull/32604#discussion_r2172385738) (and am happy to re-review), but I think the PR is fine as is now too.

fanquake commented:
Added this to `30.0`.

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -86,24 +96,35 @@ BOOST_AUTO_TEST_CASE(logging_timer)
+        std::source_location::current(),
+        std::source_location::current(),
+        std::source_location::current(),
+    };
```
> Lastly, do we really need the RemovePrefix here, given that we know the exact file structiure?

Yup, the file name may differ across platforms and the test will fail on some platforms otherwise. Necessary because it's done here: https://github.com/bitcoin/bitcoin/blob/68ca13e1f96a9dae997558a1d7a873b10056091b/src/logging.cpp#L377

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -276,4 +297,126 @@ BOOST_FIXTURE_TEST_CASE(logging_Conf, LogSetup)
+    // Use GetFileSize because fs::file_size may require a flush to be accurate.
+    std::streamsize log_file_size{static_cast<std::streamsize>(GetFileSize(log_path))};
+
+    // Logging 1 MiB should be allowed.
```
Now uses a parameter, but I have kept the comment. I think it is clear as is? I think a follow-up could address this?

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -364,7 +366,9 @@ std::string BCLog::Logger::GetLogPrefix(BCLog::LogFlags category, BCLog::Level l
 static size_t MemUsage(const BCLog::Logger::BufferedLog& buflog)
 {
-    return buflog.str.size() + buflog.logging_function.size() + buflog.source_file.size() + buflog.threadname.size() + memusage::MallocUsage(sizeof(memusage::list_node<BCLog::Logger::BufferedLog>));
+    return memusage::DynamicUsage(buflog.str) +
```
Modified the commit message and have checked that `MemUsage` returns an accurate accounting of the memory used. Thanks for this comment, it encouraged me to verify this.

Crypt-iQ commented on test/functional/feature_config_args.py:
```diff
@@ -82,7 +82,7 @@ def test_negated_config(self):
         self.log.debug('Verifying that disabling of the config file means garbage inside of it does ' \
             'not prevent the node from starting, and message about existing config file is logged')
-        ignored_file_message = [f'[InitConfig] Data directory "{self.nodes[0].datadir_path}" contains a "bitcoin.conf" file which is explicitly ignored using -noconf.']
+        ignored_file_message = [f'Data directory "{self.nodes[0].datadir_path}" contains a "bitcoin.conf" file which is explicitly ignored using -noconf.']
```
Modified the commit message 👍

l0rinc commented on src/test/logging_tests.cpp:
```diff
@@ -276,4 +297,126 @@ BOOST_FIXTURE_TEST_CASE(logging_Conf, LogSetup)
+void LogFromLocation(int location, std::string message)
+{
+    switch (location) {
+    case 0:
+        LogInfo("%s\n", message);
+        break;
+    case 1:
+        LogInfo("%s\n", message);
+        break;
```
I doubt anyone will review it if we push it separately - but I think we could make these tests less awkward. Not a blocker from my side, just a preference

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -492,6 +534,40 @@ void BCLog::Logger::ShrinkDebugFile()
+            str.insert(0, strprintf("Restarting logging from %s:%d (%s): "
+                                    "(%d MiB) were dropped during the last hour.\n",
+                                    source_loc.file_name(), source_loc.line(), source_loc.function_name(),
+                                    dropped_bytes / (1024 * 1024)));
```
Changed, I think tracking bytes is better.

Crypt-iQ commented on src/logging.h:
```diff
@@ -82,15 +104,74 @@ namespace BCLog {
             std::chrono::seconds mocktime;
-            std::string str, logging_function, source_file, threadname;
-            int source_line;
+            std::string str, threadname;
```
Did not implement as this gave me a compilation error?

stickies-v replied:
Sorry, please disregard. I think I misread this as `std::string, threadname` and probably didn't even try compiling the suggestion.

Crypt-iQ commented:
The latest push 34fc54d5476af464407c8de64c58897a35dce5cb should address the most pressing comments. IMO there is still a bit of cleanup to be done in the tests, but I would like to leave that for a follow-up. I'm going to test this branch more, so I've updated the branch mentioned in OP that can trigger the rate limiting logic: https://github.com/Crypt-iQ/bitcoin/tree/log_ratelimiting_05192025_rpc

Windows CI failure is unrelated and is addressed by https://github.com/bitcoin/bitcoin/pull/32859

dergoegge reviewed (approved):
Code review ACK 6a7147358c9d6e3883dcdbbee9fb2c1cb4baf5ff

stickies-v reviewed (approved):
ACK 6a7147358c9d6e3883dcdbbee9fb2c1cb4baf5ff

Keeping nits to a minimum because I've already been quite loud on this PR and I want to make sure we get the improvements merged soon. Left a few nits/test comments that can be done in a follow-up (although I'd be happy to quickly re-ACK here).

Crypt-iQ commented:
> Left a few nits/test comments that can be done in a follow-up

I will make a follow-up PR post-merge to address the comments (which are all reasonable, just want to limit re-review)

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -276,4 +294,182 @@ BOOST_FIXTURE_TEST_CASE(logging_Conf, LogSetup)
+
+    BOOST_CHECK_THROW(LogFromLocationAndExpect(1, log_message, "Restarting logging"), std::runtime_error);
+
+    // Attempt to log 1MiB from location 2 and 1MiB from location 3. These exempt locations should be allowed to log
+    // without limit.
+    log_file_size = GetFileSize(log_path);
+    for (int i = 0; i < 1024; ++i) {
+        BOOST_CHECK_THROW(LogFromLocationAndExpect(2, log_message, "Excessive logging detected"), std::runtime_error);
+    }
+
+    BOOST_CHECK_MESSAGE(log_file_size < GetFileSize(log_path), "location 2 should be exempt from rate limiting");
+
+    log_file_size = GetFileSize(log_path);
+    for (int i = 0; i < 1024; ++i) {
+        BOOST_CHECK_THROW(LogFromLocationAndExpect(3, log_message, "Excessive logging detected"), std::runtime_error);
+    }
+
+    BOOST_CHECK_MESSAGE(log_file_size < GetFileSize(log_path), "location 3 should be exempt from rate limiting");
```
Will address in a follow-up, I think this is from an older version of the PR where the functions used in location 2/3 were not caught by the rate limit.

l0rinc replied:
Ok, if you're fine with that, I'm fine with that, let's do the remaining ones in follow-ups.

l0rinc reviewed (changes requested):
We're getting closer, but the PR is not ready yet. I went over the code multiple times, left nits and comments and questions all over, regardless of the severity. I hope you can untangle the different directions, I know that so many comments can be overwhelming, I hope you take it the way that I meant them.

My main concerns are that the code doesn't seem finished yet, it still contains commented out code, typos, redundant comments, dangerous recursion, prefixes for non-dropped log lines, duplicate state management - and I'm not even entirely sure we should be restricting logs per line in the first place.
Also, having so many remaining inconsistencies and mixed styles and undocumented decisions makes it harder to find code smells that would help us catch mistakes (e.g. like the ones I mentioned above).

stickies-v commented on src/logging.h:
```diff
@@ -82,15 +104,88 @@ namespace BCLog {
+    public:
+        using SchedulerFunction = std::function<void(std::function<void()>, std::chrono::milliseconds)>;
+        /**
+         * @param scheduler_func    Callable object used to schedule resetting the window. The first
```
This is a public class, explaining the parameters and interface seems very desirable to me. `scheduler_func` could mean many things, when I review code I like to be able to quickly see what something's intent is. This seems like a pretty obvious best practice?

stickies-v commented on src/logging.h:
```diff
@@ -82,15 +104,88 @@ namespace BCLog {
+        //! Interval after which the window is reset.
+        const std::chrono::seconds m_reset_window;
+        //! Suppression status of a source log location.
+        enum class Status {
+            UNSUPPRESSED,     // string fits within the limit
+            NEWLY_SUPPRESSED, // suppression has started since this string
+            STILL_SUPPRESSED, // suppression is still ongoing
+        };
```
> I'm not sure I fully understand yet why we need these states

It's because the callsites behave differently whether a log line is un, newly or still suppressed. Previously, we encapsulated all that logic inside the `Consume()` function (then called `NeedsRateLimiting()`, but that gave the function too many responsibilities (as I believe you also pointed out), so this cleans that up. It think perhaps working out the code you'd prefer would be helpful.

stickies-v commented on src/logging.h:
```diff
@@ -82,15 +104,88 @@ namespace BCLog {
+        };
+        //! Consumes `source_loc`'s available bytes corresponding to the size of the (formatted)
+        //! `str` and returns its status.
+        [[nodiscard]] Status Consume(
```
> maybe it's a code smell

Perhaps you can implement an alternative and share it here?

stickies-v commented on src/logging.cpp:
```diff
@@ -367,6 +367,30 @@ static size_t MemUsage(const BCLog::Logger::BufferedLog& buflog)
+    auto& counter{m_source_locations.try_emplace(source_loc, m_max_bytes).first->second};
+    Status status{counter.GetDroppedBytes() > 0 ? Status::STILL_SUPPRESSED : Status::UNSUPPRESSED};
+
+    if (!counter.Consume(str.size()) && status == Status::UNSUPPRESSED) {
```
> if the only production usage is inverted, we should likely invert the method's return value to make it more intuitive.

I disagree. Returning `true` on successful operation is intuitive to me.

stickies-v reviewed (commented):
> My main concerns are that the code doesn't seem finished yet, it still contains commented out code, typos, redundant comments, dangerous recursion, prefixes for non-dropped log lines, duplicate state management - and I'm not even entirely sure we should be restricting logs per line in the first place.

I've responded to most of your concerns in-line. I think you've left some excellent suggestions for a follow-up, but I've not seen anything that imo warrants not merging this PR in this current shape. I understand that you'd like to see improvements, and I'd be happy to review the follow-up (as well as re-review here if the author decides to address anything), but without blockers I think it's absolutely fair for the author to decide that these can be done in a follow-up. I'd prefer not having a third version of this PR.

stickies-v reviewed (approved):
re-ACK f47e2ea9137c3a832e07d6dd845c55d35d533fa9

No changes since 6a7147358c9d6e3883dcdbbee9fb2c1cb4baf5ff except commit message and release notes improvements, and removing a commented-out test line.

Crypt-iQ commented on src/logging.h:
```diff
@@ -31,6 +35,24 @@ extern const char * const DEFAULT_DEBUGLOGFILE;
+    {
+        // Use CSipHasher(0, 0) as a simple way to get uniform distribution.
+        return static_cast<size_t>(CSipHasher(0, 0)
+                                       .Write(std::hash<std::string_view>{}(s.file_name()))
```
Hmm, I did not think about that. The `std::hash` might not be necessary then.

Crypt-iQ commented on src/logging.h:
```diff
@@ -82,15 +104,88 @@ namespace BCLog {
+        mutable StdMutex m_mutex;
+
+        //! Counters for each source location that has attempted to log something.
+        std::unordered_map<std::source_location, LogLimitStats, SourceLocationHasher, SourceLocationEqual> m_source_locations GUARDED_BY(m_mutex);
```
I had considered it, but did not implement or bench the difference. Replacing `SourceLocationEqual` & `SourceLocationHasher` with `SourceLocationLess` could be a good future improvement.

Crypt-iQ commented:
The latest push f47e2ea9137c3a832e07d6dd845c55d35d533fa9 has a few (minor) changes:
- removing the comment in `logging_filesize_rate_limit`
- documents restart behavior in the commit message of 6e47dad25c5a167cd171e5c11d51a7bee7c8d3c6
- fixes the typo of `-logsourcelocations` in the release notes
- makes the release notes more clear wrt. `LogPrintLevel`

> We're getting closer, but the PR is not ready yet. I went over the code multiple times, left nits and comments and questions all over, regardless of the severity.

I have responded to some of the comments that were about behavior changes. I hope this isn't taken as lip-service to get the PR merged, but I will make a follow-up PR post-merge to address the style and test comments (I have a list of the comments that need to be addressed). I think the most pressing comments are about behavior changes and I have attempted to address those. I think this behavior change is overdue for the codebase and it has went through iterations across several previous PRs of striking a balance between getting this mitigation shipped into production and some other less-pressing issues.

> I know that so many comments can be overwhelming, I hope you take it the way that I meant them.

I don't take it personally, review is a scarce resource and more review is better than less review. As a newcomer, I'm realizing the (appropriate) high bar for review in this project. At this point, though, I think this PR is finished as far as behavior goes and I think it is reasonable to leave the non-behavior changes to a follow-up PR.

l0rinc commented on src/logging.h:
```diff
@@ -31,6 +35,24 @@ extern const char * const DEFAULT_DEBUGLOGFILE;
+struct SourceLocationEqual {
+    bool operator()(const std::source_location& lhs, const std::source_location& rhs) const noexcept
+    {
+        return lhs.line() == rhs.line() && std::string_view(lhs.file_name()) == std::string_view(rhs.file_name());
```
The PR description still states:
> Both the quota and time window are configurable

As far as I can tell that's not the case anymore.

> When logging is restarted a tally of how many messages/bytes were dropped is printed

Do we also print the message count?

----

> But if we stay with the per-line logging-quota, I think this should at least be explained in the PR description

I meant the reason for choosing to restrict based on source lines and not globally. I'm still not fully convinced that it's more dangerous to restrict all logging, but I can accept if others think it is - I'm probably just not up-to-date with these attacks.

l0rinc commented on src/logging.h:
```diff
@@ -82,15 +104,88 @@ namespace BCLog {
+
+        //! Counters for each source location that has attempted to log something.
+        std::unordered_map<std::source_location, LogLimitStats, SourceLocationHasher, SourceLocationEqual> m_source_locations GUARDED_BY(m_mutex);
+        //! True if at least one log location is suppressed. Cached view on m_source_locations for performance reasons.
```
Fair points, maybe we could hint at that in the first message which announces that some logs are being suppressed (explaining what the new `*` prefix indicates).

l0rinc reviewed (approved):
While there are still more nits left than I'm comfortable with, I understand wanting to have some progress - and the current PR is indeed an improvement.

Lightly tested code review ACK f47e2ea9137c3a832e07d6dd845c55d35d533fa9

I'm looking forward to ironing out the wrinkles in follow-up PRs.

dergoegge reviewed (approved):
Code review ACK f47e2ea9137c3a832e07d6dd845c55d35d533fa9

glozow commented on doc/release-notes-32604.md:
```diff
@@ -0,0 +1,8 @@
+-------
+Unconditional logging to disk is now rate limited via `LogPrintf`, `LogInfo`,
+`LogWarning, `LogError`, and the corresponding `LogPrintLevel` calls by giving
+each source location a quota of 1MiB per hour. (#32604)
```
Maybe worth adding that "All logs will be prefixed with [*] if there is at least one source location that is currently being suppressed." Somebody who sees a [*] may wonder what that means, and hopefully find that in the release notes.

glozow replied:
fwiw I didn't mean for this to hold up the PR, but thanks for taking the suggestion!

glozow reviewed (commented):
lightly reviewed ACK f47e2ea9137c3a832e07d6dd845c55d35d533fa9

maflcko reviewed (commented):
left some style nits, which are trivial to ignore, or follow-up with. Nothing blocking.

review ACK f47e2ea9137c3a832e07d6dd845c55d35d533fa9 🌬

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK f47e2ea9137c3a832e07d6dd845c55d35d533fa9 🌬
+pM3Ddrk7WibLtn6aVjOQrHL+AGn6QroR8oMDjJ+OI61/kG1FDIzwCo6dTnDvuZcMfHPBic+N1tQI2C4dPYwCg==
```

</details>

Crypt-iQ commented on src/test/util/logging.h:
```diff
@@ -33,7 +33,9 @@ class DebugLogHelper
     explicit DebugLogHelper(std::string message, MatchFn match = [](const std::string*){ return true; });
-    ~DebugLogHelper() { check_found(); }
+
+    //! Mark as noexcept(false) to catch any thrown exceptions.
```
Will address in follow-up

stickies-v commented on doc/release-notes-32604.md:
```diff
@@ -0,0 +1,10 @@
+-------
+Unconditional logging to disk is now rate limited by giving each source location
+a quota of 1MiB per hour. Unconditional logging is any logging with a log level
+higher than debug, that is `info`, `warning`, and `error`. All logs will be
```
nit: Conditional logs (or rather, everything with `should_ratelimit=false`) don't get the prefix. Perhaps the better/easier UX is to just apply it to all logs?

```suggestion
higher than debug, that is `info`, `warning`, and `error`. All unconditional logs will be
```

stickies-v reviewed (approved):
re-ACK 4c772cbd83, no changes except release notes update

glozow commented:
reACK 4c772cb

davidgumberg commented:
Post-merge ACK https://github.com/bitcoin/bitcoin/commit/4c772cbd83e502a1339e8993d192ea6416ecd45c

Lightly reviewed code, tested by running IBD with `debug=all` and `-nodebug`, no messages suppressed, ran logging related benchmarks and measured performance impact is minimal.
	
One more really minor and feel-free-to-disregard nit for the follow-up pile:

The `LOG_TIME*` macros incorrectly report source locations from `logging/timer.h` rather than where the macros are invoked. There are currently only two places where these log unconditionally, when [flushing to anchors.dat](https://github.com/bitcoin/bitcoin/blob/2d59977601ea18f189991fa027364e69293c61a9/src/addrdb.cpp#L230) and when [writing UTXO snapshots](https://github.com/bitcoin/bitcoin/blob/2d59977601ea18f189991fa027364e69293c61a9/src/rpc/blockchain.cpp#L3182-L3184) so this is not a problem today AFAICT, but it is a bit of a footgun for future users of `LOG_TIME*`, one approach to solving it could look like: https://github.com/davidgumberg/bitcoin/commit/5fb07e80a422cf7b65da94dda2e1c0619c80eab3

<details>

<summary> 

#### Benchmark results

</summary>

```console
$ ./build/bin/bench_bitcoin -filter="Log.*" -min-time=10000
```

#### Branch (4c772cbd83e502a1339e8993d192ea6416ecd45c)

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|              901.02 |        1,109,855.48 |    0.7% |       14,673.87 |        3,198.96 |  4.587 |       3,030.13 |    0.1% |     10.83 | `LogWithDebug`
|              903.56 |        1,106,735.32 |    0.4% |       15,262.89 |        3,206.54 |  4.760 |       3,119.37 |    0.1% |     10.81 | `LogWithThreadNames`
|                2.61 |      383,423,939.51 |    0.1% |           70.00 |           11.18 |  6.263 |          16.00 |    0.0% |     10.99 | `LogWithoutDebug`
|              817.98 |        1,222,522.02 |    0.6% |       13,723.71 |        2,866.27 |  4.788 |       2,793.33 |    0.1% |     10.77 | `LogWithoutThreadNames`
|               28.37 |       35,246,387.38 |    0.7% |          440.00 |          121.69 |  3.616 |         107.00 |    0.0% |     11.04 | `LogWithoutWriteToFile`

#### Pre-merge master (b7e9dc8e46)

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|              883.86 |        1,131,404.74 |    1.1% |       14,673.94 |        3,136.38 |  4.679 |       3,030.14 |    0.1% |     10.83 | `LogWithDebug`
|              897.16 |        1,114,625.39 |    0.7% |       15,262.84 |        3,189.17 |  4.786 |       3,119.36 |    0.1% |     10.84 | `LogWithThreadNames`
|                2.60 |      384,716,694.70 |    0.3% |           70.00 |           11.15 |  6.279 |          16.00 |    0.0% |     11.01 | `LogWithoutDebug`
|              816.35 |        1,224,958.62 |    0.2% |       13,723.80 |        2,852.35 |  4.811 |       2,793.35 |    0.1% |     10.72 | `LogWithoutThreadNames`
|               28.28 |       35,357,376.60 |    0.4% |          440.00 |          121.30 |  3.627 |         107.00 |    0.0% |     11.01 | `LogWithoutWriteToFile`

</details>

fanquake commented:
Backported in #33225.

Crypt-iQ commented on src/test/util/logging.h:
```diff
@@ -33,7 +33,9 @@ class DebugLogHelper
-    ~DebugLogHelper() { check_found(); }
+
+    //! Mark as noexcept(false) to catch any thrown exceptions.
+    ~DebugLogHelper() noexcept(false) { check_found(); }
```
I think this can be removed now, it was added to be used with `BOOST_CHECK_THROW` / `BOOST_CHECK_NO_THROW` which are no longer used after https://github.com/bitcoin/bitcoin/pull/33011.

vasild replied:
Done in https://github.com/bitcoin/bitcoin/pull/33388

mkenshin1688-collab replied:
> Note that exception-inside-another-exception will still terminate the program, regardless of the added `noexcept(false)`. Or this:
> 
> ```cpp
> class A
> {
> public:
>     ~A() noexcept(false)
>     {
>         throw 5;
>     }
> };
> 
> int main(int, char**)
> {
>     try {
>         std::vector<A> a{4};
>     } catch (...) {
>         std::cout << "catch\n";
>     }
> 
>     return 0;
> }
> ```

gmaxwell commented:
Is there some way of disabling this that I'm missing (obviously I read the code and don't see one).  Just lost a couple days of diagnosis by this hitting exactly the event I was watching for, and also the surprise [*] made a mess of all my log processing. :(

l0rinc commented:
[-nologratelimit](https://github.com/bitcoin/bitcoin/blob/ddf2a064deb2df91f92c8aefa44437303165169c/src/init/common.cpp#L41) was added in https://github.com/bitcoin/bitcoin/pull/33011 as DEBUG_ONLY

Crypt-iQ commented:
We could make the option not DEBUG_ONLY, but I pushed back on that because I was worried users would just turn off the flag and potentially have their disk fill up slowly depending on their setup

darosior commented:
I think this is fine to expose: a user can already set `-debug` to any verbose category and fill their disk up.

l0rinc commented:
> I think this is fine to expose

Done in https://github.com/bitcoin/bitcoin/pull/35604
